frontend/src/metabase/visualizations/visualizations/PivotTable.jsx

```diff
@@ -196,6 +194,151 @@ export default class PivotTable extends Component {
       );
     }
 
+    // Create three memoized cell renderers
+    // These are tied to the `multiLevelPivot` call, so they're awkwardly shoved in render for now
+
+    const topHeaderRenderer = _.memoize(
```

Comment:
I do find that pulling these out makes it a bit easier to parse what's what. 👍

Reply:
I think longer term they shouldn't be in render at all, but this is better than the very nested JSX I had before.

Reply:
Do we have any control over the cache lifetime of memoize?
Will that be an issue or will memoize automatically take care of it (stale items got purged etc)?

Reply:
All the memoization happens within the scope of a single PivotTable render call. That removes any risk of staleness, but leaves a lot of room for better performance since even unrelated changes like updating the width blow away the cache. The next step is probably only invalidating when the data changes.